packages/esm-amd-loader/src/esm-amd-loader.ts

```diff
@@ -18,42 +18,310 @@ interface Window {
   };
 }
 
-type Registry = {
-  [url: string]: Module
-};
+type ResolveCallback = (...args: Array<{}>) => void;
+type ErrorCallback = (error: Error) => void;
+type NormalizedUrl = string&{_normalized: never};
+
+(function() {
+
+/**
+ * Describes the state machine for loading a single module.
+ *
+ * At runtime, these are integers that are inlined in their use sites.
+ */
+const enum StateEnum {
```

Comment:
I really don't like TS enums - they're one of the parts of TS that's not just a type-system superset of JS.

Given the great support for string literal and union types, I'd rather see a pattern like:

```ts
type State = 'Initialized'|'Loading'...;

/**
 * The initial state
 */
const initialized = 'Initialized' as 'Initialized';
```

This also lets you create other groups of states, like:

```ts
type WaitingState = 'WaitingOnEarlierScripts'|'WaitingOnDeps';
```

Reply:
That's normally what I would lean towards too, but this form keeps about the same ergonomics and legibility and compiles down to a single byte. If we want to convert this to JS we could maintain this by just declaring a const for each enum value, and we could count in them being inlined as they are now.

Reply:
I assumed this was more about size over the wire?